### PR TITLE
Fix S2225 FP: Support static virtual/abstract interface methods

### LINKED ISSUE #6307: Fix S2225 FP: Support static virtual/abstract interface methods
### Description

FP for S2225 (ToStringShouldNotReturnNull).

Using C#11 static virtual/abstract interface methods, the rule is erroneously raising an issue.

### Repro steps

```csharp
public interface SomeInterface
{
    static virtual string ToString()
    {
        return null; // Noncompliant FP
    }
}

public interface SomeOtherInterface
{
    static abstract string ToString();
}

public class SomeClass: SomeOtherInterface
{
    public static string ToString()
    {
        return null; // Noncompliant FP
    }
}

```

### Expected behavior

The rule should not raise an issue.

### Actual behavior

The rule does raise an issue.

### PR DESCRIPTION
**Pull request by cristian-ambrosini-sonarsource**
Fixes #6307

### CODE REVIEW

**Review by gregory-paidis-sonarsource — APPROVED** [review]

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ToStringShouldNotReturnNull.CSharp11.cs [hunk 1]
```diff
@@ -2,7 +2,16 @@
 {
     static virtual string ToString()
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ToStringShouldNotReturnNull.CSharp11.cs:3]: Please also add a test case with a static ToString method in a class that does not come from an interface. (Add that to the base test cases)

File: analyzers/src/SonarAnalyzer.VisualBasic/Rules/ToStringShouldNotReturnNull.cs [hunk 2]
```diff
@@ -42,4 +42,6 @@ public sealed class ToStringShouldNotReturnNull : ToStringShouldNotReturnNullBas
 
     protected override bool IsLocalOrLambda(SyntaxNode node) =>
         node.IsKind(SyntaxKind.MultiLineFunctionLambdaExpression);
+
+    protected override bool IsStatic(SyntaxNode node) => false;
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.VisualBasic/Rules/ToStringShouldNotReturnNull.cs:46]: Actually, for VB you can have "Static" members which are called Shared in VB.NET. You can use [this ](https://converter.telerik.com/)page to convert code samples from C# to VB.NET.

File: analyzers/src/SonarAnalyzer.Common/Rules/ToStringShouldNotReturnNullBase.cs [hunk 3]
```diff
@@ -63,5 +65,6 @@ protected void ToStringReturnsNull(SyntaxNodeAnalysisContext context, SyntaxNode
         node.Ancestors()
             .TakeWhile(x => !IsLocalOrLambda(x))
             .Any(x => Language.Syntax.IsKind(x, MethodKind)
-                && nameof(ToString).Equals(Language.Syntax.NodeIdentifier(x)?.ValueText, Language.NameComparison));
+                && nameof(ToString).Equals(Language.Syntax.NodeIdentifier(x)?.ValueText, Language.NameComparison)
+                && !IsStatic(x));
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/ToStringShouldNotReturnNullBase.cs:69]: As the case makes sense for both C# and VB.NET. I suggest to add a helper method also for VB.NET to check if a sub or function is static/shared. And after add a method to Language.Syntax (IsStatic) that will call the checks for C# and VB.NET and will return true if a method or sub or function is static/shared. This way you should be able to do the call from here and you would not need the abstract IsStatic method.

**Review by csaba-sagi-sonarsource — CHANGES_REQUESTED** [review]
I left a couple of comments to support also VB.NET

**Comment by csaba-sagi-sonarsource** [comment]
Please address the  code smells.

**Comment by csaba-sagi-sonarsource** [comment]
Please add a test case for VB.NET with a Shared Function called ToString that returns String.

File: analyzers/src/SonarAnalyzer.CSharp/Facade/CSharpSyntaxFacade.cs [hunk 4]
```diff
@@ -122,5 +122,11 @@ public override string NodeStringTextValue(SyntaxNode node, SemanticModel semant
 
         public override bool TryGetGetInterpolatedTextValue(SyntaxNode node, SemanticModel semanticModel, out string interpolatedValue) =>
             Cast<InterpolatedStringExpressionSyntax>(node).TryGetGetInterpolatedTextValue(semanticModel, out interpolatedValue);
+
+        public override bool IsStatic(SyntaxNode node)
+        {
+            var methodDeclaration = (MethodDeclarationSyntax)node;
+            return methodDeclaration.IsStatic();
+        }
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Facade/CSharpSyntaxFacade.cs:130]:
```suggestion
        public override bool IsStatic(SyntaxNode node) =>
            Cast<BaseMethodDeclarationSyntax>(node).IsStatic();
```

File: analyzers/src/SonarAnalyzer.Common/Rules/ToStringShouldNotReturnNullBase.cs [hunk 5]
```diff
@@ -63,5 +63,9 @@ protected void ToStringReturnsNull(SyntaxNodeAnalysisContext context, SyntaxNode
...
+
+    private bool IsStatic(SyntaxNode node) =>
+        Language.Syntax.IsStatic(node);
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/ToStringShouldNotReturnNullBase.cs:70]: Why have this method when you can call directly the method from the SyntaxFacade?

File: analyzers/src/SonarAnalyzer.VisualBasic/Extensions/SyntaxNodeExtensions.cs [hunk 6]
```diff
@@ -81,6 +81,9 @@ static bool TakesExpressionTree(SymbolInfo info)
             }
         }
 
+        public static bool IsShared(this SyntaxNode node) =>
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.VisualBasic/Extensions/SyntaxNodeExtensions.cs:84]: You will need a new class and the proper extension to check if the sub or function is shared:

```
internal static class MethodBlockSyntaxExtensions
{
    public static bool IsShared(this MethodBlockSyntax methodBlock) =>
        methodBlock.SubOrFunctionStatement.Modifiers.Any(SyntaxKind.SharedKeyword);
}
```

File: analyzers/src/SonarAnalyzer.VisualBasic/Facade/VisualBasicSyntaxFacade.cs [hunk 7]
```diff
@@ -123,5 +123,8 @@ public override SyntaxNode RemoveConditionalAccess(SyntaxNode node)
 
         public override bool TryGetGetInterpolatedTextValue(SyntaxNode node, SemanticModel semanticModel, out string interpolatedValue) =>
             Cast<InterpolatedStringExpressionSyntax>(node).TryGetGetInterpolatedTextValue(semanticModel, out interpolatedValue);
+
+        public override bool IsStatic(SyntaxNode node) =>
+            node.IsShared();
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.VisualBasic/Facade/VisualBasicSyntaxFacade.cs:128]:
```suggestion
        public override bool IsStatic(SyntaxNode node) =>
            Cast<MethodBlockSyntax>(node).IsShared();
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ToStringShouldNotReturnNull.CSharp11.cs [hunk 8]
```diff
@@ -2,7 +2,16 @@
...
+        if (true)
+        {
+            return null; // Compliant
+        }
+        else
+        {
+            return string.Empty;
+        }
+
+        return null; // Compliant
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ToStringShouldNotReturnNull.CSharp11.cs:14]: Educational: Why did you change the test case?

**Review by csaba-sagi-sonarsource — CHANGES_REQUESTED** [review]
I left a couple of comments.

File: analyzers/src/SonarAnalyzer.VisualBasic/Extensions/MethodBlockSyntaxExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+using System.Linq;
+using Microsoft.CodeAnalysis.VisualBasic;
+using Microsoft.CodeAnalysis.VisualBasic.Syntax;
+
+namespace SonarAnalyzer.Extensions;
+
+public static class MethodBlockSyntaxExtensions
+{
+    public static bool IsShared(this MethodBlockSyntax methodBlock) =>
+        methodBlock.SubOrFunctionStatement.Modifiers.Any(x => x.Kind() == SyntaxKind.SharedKeyword);
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.VisualBasic/Extensions/MethodBlockSyntaxExtensions.cs:30]:
```suggestion
        methodBlock.SubOrFunctionStatement.Modifiers.Any(SyntaxKind.SharedKeyword);
```

**Review by csaba-sagi-sonarsource — CHANGES_REQUESTED** [review]
One last change requested.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ToStringShouldNotReturnNull.vb [hunk 10]
```diff
@@ -63,22 +63,30 @@ Namespace Compliant
...
+
+        Public Function SomeMethod() As String
             Return If(Nothing, Nothing) ' Not supported, doesn't make sense
         End Function
-	
-	End Class
+
+    End Class
+
+    Class ToStringSharedMethod
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ToStringShouldNotReturnNull.vb:83]: There is still a missing UT that is not raising because the function is shared.
This example would not raise even if it would not be shared as it returns String.Empty instead of Nothing.

File: analyzers/src/SonarAnalyzer.VisualBasic/Extensions/MethodBlockSyntaxExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Linq;
+using Microsoft.CodeAnalysis.VisualBasic;
+using Microsoft.CodeAnalysis.VisualBasic.Syntax;
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.VisualBasic/Extensions/MethodBlockSyntaxExtensions.cs:23]: Unused usings.

**Review by csaba-sagi-sonarsource — APPROVED** [review]
LGTM!
